% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Rainbow_Plot.R
\name{rainbowPlot}
\alias{rainbowPlot}
\title{Rainbow Plot.}
\usage{
rainbowPlot(
  data,
  nloci = 400,
  minWt = 0.01,
  maxFreq = 0.7,
  minSamples = 10,
  filterHard = TRUE,
  seed,
  title = "",
  printout = TRUE
)
}
\arguments{
\item{data}{A data.frame with at least the following columns.
\describe{
  \item{AltProp}{A numberic vector giving the proportion (of reads mapping to the exogenous genome) that carry the non-standard alleles thought to be in the vagrant copies}
  \item{Position}{A factor (or structure that can be coerced to a factor),
  giving a unique name for each SNP location.}
  \item{Sample}{A factor (or structure that can be coerced to a factor),
  giving a unique name for each sample.}
  \item{ylog}{A numeric vector giving the log(AltProp)}
  \item{xnqlogis}{A numeric vector giving log(m/N);
  where m is the number of reads mapping to the exogenous genome and
  N is the remaining reads.}
  }}

\item{nloci}{The number of loci to be selected for the analysis. Default, 400.}

\item{minWt}{The minimum average allele frequency of SNPs to be included in the analysis.
Default, 0.01.}

\item{maxFreq}{The maximum allele frequency of an individual observation to be included
in the analysis. Default, 0.7.}

\item{minSamples}{The minimum number of samples in which a SNP should be called in order to
be included in the analysis.}

\item{filterHard}{If filterHard is TRUE, the SNP loci with slopes in the outer quartiles are
discarded. Otherwise the outliers identified by the default method of boxplot.stats function
are discarded. Default, TRUE.}

\item{seed}{Random number seed.}

\item{title}{User-supplied title for the rainbow plot.}

\item{printout}{If printout is TRUE, the function prints the estimates. Default, TRUE.}
}
\value{
An invisible list with the following elements.
\describe{
 \item{$intercepts}{A vector giving the intercept estimate,
 the lower and upper 95\% confidence interval bounds.}
 \item{$depth.est}{A crude upper limit; the minimum over samples
 of the proportion of reads mapping to the exogenous genome.
 Some part (or all) of the CI would be expected to be below this value.}
 \item{$num.loci}{The number of loci remaining after filtering,
 which were used to obtain the intercept estimates.}
 \item{$lmer.model}{An lmer object which contains the details of the mixed effects model fitted to identify
 the largest intercept and its standard error.
 In the case of unusual or problematic results it may be useful to inspect the residuals, see examples.}
 \item{$func.params}{A record of the function call}
}
}
\description{
Generate a plot to estimate the proportion of the nuclear genome
made up of a particular vagrant sequence. The function produces the plots
intercept estimate, and mapping depth estimate described by
Becher & Nichols (2021) citation.
}
\examples{
## Access one of the package's example data-sets (hopperDF, parrotDF or humanDF)
data(hopperDF)
##
## plot and printout (by default).
rainbowPlot(hopperDF, seed = 12345, title = "Grasshopper")
##
## plot without printing the results and store results in res1.
res1 <- rainbowPlot(hopperDF, seed = 12345, printout = False, title = "Grasshopper")
## print just the stored estimates (the first two elements of the list)
print(res1[1:2])
## Inspect the residuals of the lmer model
plot(res1$lmer.model)
}
